cs/ccxt/base/Throttler.cs

```diff
@@ -73,23 +89,85 @@ private async Task loop()
+                }
+            }
+            (Task, double) first;
+            lock (queueLock)
+            {
+                first = this.queue.Peek();
+            }
+            var task = first.Item1;
+            var cost = first.Item2;
+            var now = milliseconds();
+            timestamps = timestamps.Where(t => now - t.timestamp < Convert.ToDouble(this.config["windowSize"])).ToList();
+            var totalCost = timestamps.Sum(t => t.cost);
+            if (totalCost + cost <= Convert.ToDouble(this.config["maxWeight"]))
+            {
+                timestamps.Add((now, cost));
+                await Task.Delay(0);
+                if (task != null && task.Status == TaskStatus.Created)
+                {
+                    task.Start();
+                }
+                lock (queueLock)
+                {
+                    this.queue.Dequeue();
+                    if (this.queue.Count == 0) this.running = false;
+                }
+            }
+            else
+            {
+                var earliest = timestamps[0].timestamp;
+                var waitTime = (earliest + Convert.ToDouble(this.config["windowSize"])) - now;
+                if (waitTime > 0)
+                {
+                    await Task.Delay((int)waitTime);
+                }
+            }
+        }
+    }
 
-        var cost = (cost2 != null) ? Convert.ToDouble(cost2) : Convert.ToDouble(this.config["cost"]);
-        if (this.queue.Count > (int)this.config["maxCapacity"])
+    private async Task loop()
+    {
+        if (this.config["algorithm"] as string == "leakyBucket")
         {
-            throw new Exception("throttle queue is over maxCapacity (" + this.config["maxCapacity"].ToString() + "), see https://github.com/ccxt/ccxt/issues/11645#issuecomment-1195695526");
+            await leakyBucketLoop();
         }
-        var t = new Task(() => { });
-        this.queue.Enqueue((t, cost));
-        if (!this.running)
+        else
         {
-            this.running = true;
-            // Task.Run(() => { this.loop(); });
-            this.loop();
+            await rollingWindowLoop();
+        }
+    }
+
+    public async Task<Task> throttle(object cost2)
+    {
+        var cost = (cost2 != null) ? Convert.ToDouble(cost2) : Convert.ToDouble(this.config["cost"]);
+        lock (queueLock)
+        {
+            if (this.queue.Count > (int)this.config["maxLimiterRequests"])
+            {
+                throw new Exception("throttle queue is over maxLimiterRequests (" + this.config["maxLimiterRequests"].ToString() + "), see https://github.com/ccxt/ccxt/issues/11645#issuecomment-1195695526");
+            }
+            var t = new Task(() => { });
+            this.queue.Enqueue((t, cost));
+            if (!this.running)
+            {
+                this.running = true;
+                // Task.Run(() => { this.loop(); });
+                this.loop();
+            }
+            return t;
```

Comment:
@carlosmiei I think this might be a problem, we are in `lock` scope here, and `this.loop` is being called. maybe we should end `lock` several lines above where it's `this.queue.Enqueue...` ?

```suggestion
        }
        if (!this.running)
        {
            this.running = true;
            // Task.Run(() => { this.loop(); });
            this.loop();
        }
        return t;
```

Reply:
we could maybe do that, I added all the lock stuff in because I was running into some errors when I got up to and above 100 concurrent requests, I can't remember exactly which parts fixed it, but at least some of it is necessary